src/apps/lwaftr/lwaftr.lua

```diff
@@ -43,6 +44,43 @@ local function compute_binding_table_by_ipv4(binding_table)
    return ret
 end
 
+local function port_to_psid(port, reserved_ports_bit_count, psid_len)
+   local shift = 16 - (reserved_ports_bit_count + psid_len)
+   return band(rshift(port, shift), (2^psid_len)-1)
+end
+
+local function compute_psid_info_map(binding_table)
+   local value_type = ffi.typeof([[
+      struct { uint16_t reserved_ports_bit_count; uint16_t psid_len; }
+   ]])
+   local function extract_value(entry)
+      local psid_info = entry[3]
+      local value = ffi.new(value_type)
+      value.reserved_ports_bit_count = psid_info.reserved_ports_bit_count or 0
+      value.psid_len = psid_info.psid_len or 0
+      return value
+   end
+   local function extract_key_and_value(entry)
+      local ipv4 = entry[2]
+      return ipv4, extract_value(entry)
+   end
+   local builder = rangemap.RangeMapBuilder.new(value_type)
+   for _, entry in ipairs(binding_table) do
+      local key, value = extract_key_and_value(entry)
+      builder:add(key, value)
+   end
+   local map = builder:build()
+   function map:lookup_value(ipv4)
```

Comment:
I kinda think we should not allow the value to escape like this.  The reason is, the pointer is a heap object, and I am not sure that scalar replacement will do the right thing always.  That's why in a previous patch I suggested returning a and k directly (but really we should return k and m, or rather psid_len and shift), so we pretty much know that the value allocation will be removed.  Maybe it is OK though.